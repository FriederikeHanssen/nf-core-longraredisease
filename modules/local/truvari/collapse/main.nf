process TRUVARI_COLLAPSE {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "biocontainers/truvari:5.3.0--pyhdfd78af_0"

    input:
    tuple val(meta), path(vcf), path(tbi), path(bed)
    val(refdist)     // Reference distance for merging
    val(pctsim)      // Percent similarity for merging
    val(pctseq)      // Percent sequence for merging
    val(passonly)    // Boolean: Flag to only keep PASS variants
    val(dup_to_ins)  // Boolean: Flag to treat Duplications as Insertions

    output:
    tuple val(meta), path("*_merged.vcf")         , emit: merged_vcf
    tuple val(meta), path("*_collapsed.vcf")      , emit: collapsed_vcf
    path "versions.yml"                           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    // Convert boolean values to command-line flags
    def passonly_flag = passonly ? '--passonly' : ''
    def dup_to_ins_flag = dup_to_ins ? '--dup-to-ins' : ''
    def bed_flag = (bed && bed.name != 'NO_FILE') ? "--bed ${bed}" : ''

    """
    truvari collapse \\
        -i ${vcf} \\
        -o ${prefix}_merged.vcf \\
        -r ${refdist} \\
        -P ${pctsim} \\
        --pctseq ${pctseq} \\
        ${bed_flag} \\
        ${passonly_flag} \\
        ${dup_to_ins_flag} \\
        -c ${prefix}_collapsed.vcf \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        truvari: \$(echo \$(truvari version 2>&1) | sed 's/^Truvari v//' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_merged.vcf
    touch ${prefix}_collapsed.vcf

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        truvari: \$(echo \$(truvari version 2>&1) | sed 's/^Truvari v//' )
    END_VERSIONS
    """
}