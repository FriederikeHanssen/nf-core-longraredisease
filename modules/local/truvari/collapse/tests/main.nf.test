nextflow_process {

    name "Test Process TRUVARI_COLLAPSE"
    script "../main.nf"
    process "TRUVARI_COLLAPSE"

    tag "modules"
    tag "modules_nfcore"
    tag "truvari"
    tag "truvari/collapse"

    test("without bed") {

        when {
            process {
                """
                input[0] = [
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz', checkIfExists:true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi', checkIfExists:true),
                    []
                ]
                input[1] = 1000      // refdist
                input[2] = 0.70      // pctsim
                input[3] = 0.70      // pctseq
                input[4] = false     // passonly
                input[5] = false     // dup_to_ins
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.merged_vcf[0][1]).vcf.summary,
                    path(process.out.collapsed_vcf[0][1]).vcf.summary,
                    path(process.out.versions[0]).yaml
                    ).match()
                }
            )
        }
    }

    test("with bed") {

        when {
            process {
                """
                input[0] = [
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz', checkIfExists:true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi', checkIfExists:true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.bed', checkIfExists:true)
                ]
                input[1] = 1000      // refdist
                input[2] = 0.70      // pctsim
                input[3] = 0.70      // pctseq
                input[4] = true      // passonly
                input[5] = false     // dup_to_ins
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.merged_vcf[0][1]).vcf.summary,
                    path(process.out.collapsed_vcf[0][1]).vcf.summary,
                    path(process.out.versions[0]).yaml
                    ).match()
                }
            )
        }
    }

    test("with dup_to_ins flag") {

        when {
            process {
                """
                input[0] = [
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz', checkIfExists:true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi', checkIfExists:true),
                    []
                ]
                input[1] = 500       // refdist
                input[2] = 0.80      // pctsim
                input[3] = 0.80      // pctseq
                input[4] = false     // passonly
                input[5] = true      // dup_to_ins
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.merged_vcf[0][1]).vcf.summary,
                    path(process.out.collapsed_vcf[0][1]).vcf.summary,
                    path(process.out.versions[0]).yaml
                    ).match()
                }
            )
        }
    }

    test("without bed - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz', checkIfExists:true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi', checkIfExists:true),
                    []
                ]
                input[1] = 1000      // refdist
                input[2] = 0.70      // pctsim
                input[3] = 0.70      // pctseq
                input[4] = false     // passonly
                input[5] = false     // dup_to_ins
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml
                    ).match()
                }
            )
        }
    }
}