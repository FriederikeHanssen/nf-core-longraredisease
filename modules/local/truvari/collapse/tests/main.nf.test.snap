{
    "without bed": {
        "content": [
            "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=15, phased=false, phasedAutodetect=true]",
            "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=12, phased=false, phasedAutodetect=true]",
            {
                "TRUVARI_COLLAPSE": {
                    "truvari": "5.3.0"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-01-07T14:34:37.210063"
    },
    "with bed": {
        "content": [
            "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=10, phased=false, phasedAutodetect=true]",
            "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=8, phased=false, phasedAutodetect=true]",
            {
                "TRUVARI_COLLAPSE": {
                    "truvari": "5.3.0"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-01-07T14:36:48.501008"
    },
    "with dup_to_ins flag": {
        "content": [
            "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=13, phased=false, phasedAutodetect=true]",
            "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=11, phased=false, phasedAutodetect=true]",
            {
                "TRUVARI_COLLAPSE": {
                    "truvari": "5.3.0"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-01-07T14:38:15.123456"
    },
    "without bed - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_collapsed.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,b3800d0192adcda9407236f1111be2e3"
                ],
                "collapsed_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test_collapsed.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "merged_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b3800d0192adcda9407236f1111be2e3"
                ]
            },
            {
                "TRUVARI_COLLAPSE": {
                    "truvari": "5.3.0"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-01-07T14:39:08.51064"
    }
}